Add virtual memory guide (#3538)
ROCm-Developer-Tools/HIP: docs/how-to/virtual_memory.rst

.. meta::
  :description: This chapter describes introduces Virtual Memory (VM) and shows
                how to use it in AMD HIP.
  :keywords: AMD, ROCm, HIP, CUDA, virtual memory, virtual, memory, UM, APU

.. _virtual_memory:

*****************************
HIP Virtual Memory Management
*****************************

> Managing memory management is important when creating high-performance applications in the HIP ecosystem. Both allocating and copying memory can result in bottlenecks, which can significantly impact performance.

Reviewer: Memory management is important...

Managing memory management seems redundant

> Global memory allocation in HIP uses the C language style allocation function. This works fine for simple cases but can cause problems if your memory needs change. If you need to increase the size of your memory, you'll have to allocate a second larger buffer and copy the data to it before you can free the original buffer. This increases overall memory usage and causes unnecessary ``memcpy`` calls. Another solution is to allocate a larger buffer than you initially need. However, this isn't an efficient way to handle resources and doesn't solve the issue of reallocation when the extra buffer runs out.

Reviewer 2: ```suggestion
Global memory allocation in HIP uses the C language style allocation function. This works fine for simple cases but can cause problems if your memory needs change. If you need to increase the size of your memory, you must allocate a second larger buffer and copy the data to it before you can free the original buffer. This increases overall memory usage and causes unnecessary ``memcpy`` calls. Another solution is to allocate a larger buffer than you initially need. However, this isn't an efficient way to handle resources and doesn't solve the issue of reallocation when the extra buffer runs out.
```

Reviewer 3: Fixed.

> Virtual memory management solves these memory management problems. It helps to reduce memory usages and unnecessary ``memcpy`` calls.

Reviewer 2: ```suggestion
Virtual memory management solves these memory management problems. It helps to reduce memory usage and unnecessary ``memcpy`` calls.
```

Reviewer 3: Fixed.

.. _memory_allocation_virtual_memory:
Memory Allocation
=================

> Previously, you called the ``hipMalloc`` function to allocate a chunk of memory on the device. ``hipMalloc`` returns a pointer, which is immediately usable after the call. When using virtual memory, this is separated into multiple steps using the ``hipMemCreate``, ``hipMemAddressReserve``, ``hipMemMap``, and ``hipMemSetAccess`` functions. This guide explains what these functions do and how you can use them for your needs.

Reviewer: Standard memory allocation uses the ``hipMalloc`` function to allocate a block of memory on the device. However, when using virtual memory, this process is separated into multiple steps using the ``hipMemCreate``, ``hipMemAddressReserve``, ``hipMemMap``, and ``hipMemSetAccess`` functions. This guide explains what these functions do and how you can use them for virtual memory management.

> Previously, you called the ``hipMalloc`` function to allocate a chunk of memory on the device. ``hipMalloc`` returns a pointer, which is immediately usable after the call. When using virtual memory this is separated into multiple steps using the ``hipMemCreate``, ``hipMemAddressReserve``, ``hipMemMap`` and ``hipMemSetAccess`` functions. In this guide you'll learn what these functions do and how you can use them for your needs.

Reviewer 2: ```suggestion
Previously, you called the ``hipMalloc`` function to allocate a chunk of memory on the device. ``hipMalloc`` returns a pointer, which is immediately usable after the call. When using virtual memory, this is separated into multiple steps using the ``hipMemCreate``, ``hipMemAddressReserve``, ``hipMemMap``, and ``hipMemSetAccess`` functions. This guide explains what these functions do and how you can use them for your needs.
```

Reviewer 3: Fixed.

Allocate Physical Memory
------------------------

> The first step is allocate the physical memory itself with the ``hipMemCreate`` function. This function accepts the size of the buffer, an ``unsigned long long`` variable for the flags, and a ``hipMemAllocationProp`` variable. ``hipMemAllocationProp`` contains the properties of the memory to be allocated, such as where the memory is physically located and what kind of shareable handles are available. If the allocation is successful, the function returns a value of ``hipSuccess``, with ``hipMemGenericAllocationHandle_t`` representing a valid physical memory allocation. The allocated memory size has to be aligned with the granularity appropriate for the properties of the allocation. You can use the ``hipMemGetAllocationGranularity`` function to determine the correct granularity.

Reviewer 2: ```suggestion
The first step is to allocate the physical memory itself with the ``hipMemCreate`` function. This function accepts the size of the buffer, an ``unsigned long long`` variable for the flags, and a ``hipMemAllocationProp`` variable. ``hipMemAllocationProp`` contains the properties of the memory to be allocated, such as where the memory is physically located and what kind of shareable handles are available. If the allocation is successful, the function returns a value of ``hipSuccess``, with ``hipMemGenericAllocationHandle_t`` representing a valid physical memory allocation. The allocated memory size must be aligned with the granularity appropriate for the properties of the allocation. You can use the ``hipMemGetAllocationGranularity`` function to determine the correct granularity.
```

Reviewer 3: Fixed.

.. code-block:: cpp

    size_t granularity = 0;
    hipMemGenericAllocationHandle_t allocHandle;
    hipMemAllocationProp prop = {};
    prop.type = HIP_MEM_ALLOCATION_TYPE_PINNED;
    prop.location.type = HIP_MEM_LOCATION_TYPE_DEVICE;
    prop.location.id = currentDev;
    hipMemGetAllocationGranularity(&granularity, &prop, HIP_MEM_ALLOC_GRANULARITY_MINIMUM);
    padded_size = ROUND_UP(size, granularity);
    hipMemCreate(&allocHandle, padded_size, &prop, 0);

Reserve Virtual Address Range
-----------------------------

> After you have acquired the allocation of physical memory, you must map it before you can use it. To do so, you need a virtual address to map it to.  Mapping means that the physical memory allocation is available by means of the virtual address range it is mapped to. To reserve a virtual memory range, use the ``hipMemAddressReserve`` function. The size of the virtual memory must match the amount of physical memory previously allocated. You can then map the physical memory allocation to the newly-acquired virtual memory address range using the ``hipMemMap`` function.

Reviewer 2: ```suggestion
After you have acquired an allocation of physical memory, you must map it before you can use it. To do so, you need a virtual address to map it to.  Mapping means the physical memory allocation is available from the virtual address range it is mapped to. To reserve a virtual memory range, use the ``hipMemAddressReserve`` function. The size of the virtual memory must match the amount of physical memory previously allocated. You can then map the physical memory allocation to the newly-acquired virtual memory address range using the ``hipMemMap`` function.
```

Reviewer 3: Fixed.

.. code-block:: cpp

    hipMemAddressReserve(&ptr, padded_size, 0, 0, 0);
    hipMemMap(ptr, padded_size, 0, allocHandle, 0);

Set Memory Access
-----------------

> Finally, use the ``hipMemSetAccess`` function to enable memory access. It accepts the pointer to the virtual memory, the size, and a ``hipMemAccessDesc`` descriptor as parameters.,mm

Reviewer 2: ```suggestion
Finally, use the ``hipMemSetAccess`` function to enable memory access. It accepts the pointer to the virtual memory, the size, and a ``hipMemAccessDesc`` descriptor as parameters.
```

Reviewer 3: Fixed.

> Finally, you have to enable access to the memory. This is achieved by using the ``hipMemSetAccess`` function. It accepts the pointer to the virtual memory, the size, and a ``hipMemAccessDesc`` descriptor as parameters.

Reviewer 2: ```suggestion
Finally, use the ``hipMemSetAccess`` function to enable memory access. It accepts the pointer to the virtual memory, the size, and a ``hipMemAccessDesc`` descriptor as parameters.
```

Reviewer 3: Updated.

Reviewer 2: It looks like a couple of stray characters were added after this sentence by mistake. I'll indicate the error on the actual file as well.

Reviewer 3: Oh, thanks! Fixed.

.. code-block:: cpp

    hipMemAccessDesc accessDesc = {};
    accessDesc.location.type = HIP_MEM_LOCATION_TYPE_DEVICE;
    accessDesc.location.id = currentDev;
    accessDesc.flags = HIP_MEM_ACCESS_FLAGS_PROT_READWRITE;
    hipMemSetAccess(ptr, padded_size, &accessDesc, 1);

At this point the memory is allocated, mapped, and ready for use. You can read and write to it, just like you would a C style memory allocation.

Free Virtual Memory
-------------------

To free the memory allocated in this manner, use the corresponding free functions. To unmap the memory, use ``hipMemUnmap``. To release the virtual address range, use ``hipMemAddressFree``.  Finally, to release the physical memory, use ``hipMemRelease``.

.. code-block:: cpp

    hipMemUnmap(ptr, size);
    hipMemRelease(allocHandle);
    hipMemAddressFree(ptr, size);

.. _usage_virtual_memory:

> Usage

Reviewer: Memory usage

=====

> Increase Allocation Size

Reviewer: Should this be _Dynamically increase allocation size_? It seems this is the crux of the reason for using virtual memory, isn't it?

------------------------

The ``hipMemAddressReserve`` function allows you to increase the amount of pre-allocated memory. This function accepts a parameter representing the requested starting address of the virtual memory. This allows you to have a continuous virtual address space without worrying about the underlying physical allocation.

.. code-block:: cpp

    hipMemAddressReserve(&new_ptr, (new_size - padded_size), 0, ptr + padded_size, 0);
    hipMemMap(new_ptr, (new_size - padded_size), 0, newAllocHandle, 0);
    hipMemSetAccess(new_ptr, (new_size - padded_size), &accessDesc, 1);

The code sample above assumes that ``hipMemAddressReserve`` was able to reserve the memory address at the specified location. However, this isn't guaranteed to be true, so you should validate that ``new_ptr`` points to a specific virtual address before using it.

> Avoid Synchronization

Reviewer: I think I would just make this point under the _Free virtual memory_ section above. It seems like a benefit of those functions rather than a separate feature to be discussed here.

---------------------

Another major advantage of virtual memory management is the lack of synchronization when memory is released. If you call ``hipFree`` when you have multiple streams running in parallel, it synchronizes the device. This causes worse resource usage and performance.

> Enable Peer Access

Reviewer: I think I would make this point under the section of _Set memory access_. This is just another benefit of those functions, and is something that should be done at that time as well.

.. meta::
  :description: This chapter describes introduces Virtual Memory (VM) and shows
                how to use it in AMD HIP.
  :keywords: AMD, ROCm, HIP, CUDA, virtual memory, virtual, memory, UM, APU

.. _virtual_memory:

*******************************************************************************

> HIP Virtual Memory Management

Reviewer 2: Use title case for the heading (unless this is an official name for an official process)

Reviewer 3: Fixed.

*******************************************************************************

> Managing memory management is important when creating high-performance applications in the HIP ecosystem. Both memory allocation and copy can result in bottlenecks, which can significantly impact performance. 

Reviewer 2: ```suggestion
Managing memory management is important when creating high-performance applications in the HIP ecosystem. Both allocating and copying memory can result in bottlenecks, which can significantly impact performance. 
```

Reviewer 3: Updated.

> The global memory allocation in HIP happens with the C style allocation function. This

Reviewer 2: ```suggestion
Global memory allocation in HIP uses the C language style allocation function. This
```

Reviewer 3: Updated.

> works fine for simple cases, but there is a problem, if your memory needs change. If you

Reviewer 2: ```suggestion
works fine for simple cases but can cause problems if your memory needs change. If you
```

Reviewer 3: Updated.

need to increase the size of your memory, you'll have to allocate a second larger buffer

> and copy the data to it, before you can free the original buffer, which increases overall
> memory usage and causes unnecessary ``memcpy`` calls. An other solution is to have a

Reviewer 2: ```suggestion
and copy the data to it before you can free the original buffer. This increases overall
memory usage and causes unnecessary ``memcpy`` calls. Another solution is to have a
```

> and copy the data to it, before you can free the original buffer, which increases overall
> memory usage and causes unnecessary ``memcpy`` calls. An other solution is to have a
> larger buffer allocated, than you initially need, which isn't an efficient way to handle
> resources and doesn't solve the issue of reallocation, when the extra buffer runs out.

Reviewer 2: ```suggestion
and copy the data to it before you can free the original buffer. This increases overall
memory usage and causes unnecessary ``memcpy`` calls. Another solution is to allocate a
larger buffer than you initially need. However, this isn't an efficient way to handle
resources and doesn't solve the issue of reallocation when the extra buffer runs out.
```

Reviewer 3: Updated.

> Virtual memory management is the solution to these memory management problems. It helps

Reviewer 2: ```suggestion
Virtual memory management solves these memory management problems. It helps
```

Reviewer 3: Updated.

to reduce memory usages and unnecessary ``memcpy`` calls.

.. _memory_allocation_virtual_memory:
Memory allocation
===============================================================================

> When you need a chunk of memory on the device, previously you called the ``hipMalloc``
> function to allocate and it returns a pointer, that is immediately usable after the call.

Reviewer 2: ```suggestion
Previously, you called the ``hipMalloc`` function to allocate a chunk of memory on the device. ``hipMalloc`` returns a pointer, which is immediately usable after the call.
```

Reviewer 3: Updated

When using virtual memory this is separated into multiple steps using the
``hipMemCreate``, ``hipMemAddressReserve``, ``hipMemMap`` and ``hipMemSetAccess``
functions. In this guide you'll learn what these functions do and how you can use them
for your needs.

Allocate physical memory
-------------------------------------------------------------------------------

The first step is allocate the physical memory itself with the ``hipMemCreate`` function.
It take the size of the buffer, an ``unsigned long long`` for flags and a
``hipMemAllocationProp``, which contains the properties of the properties of the of the
memory to be allocated, like where is this memory physically located or what kind of
shareable handles should be available. If the allocation is successful the function
returns with a ``hipSuccess`` and the ``hipMemGenericAllocationHandle_t`` will represent

> a valid physical memory allocation. It is important to note that the size has to be

Reviewer 2: I'm not sure what "the size" refers to here. Is this the value in ``hipMemGenericAllocationHandle_t``? Please clarify this statement.

Reviewer 3: It is related to the allocated memory size. Updated.

> It take the size of the buffer, an ``unsigned long long`` for flags and a
> ``hipMemAllocationProp``, which contains the properties of the properties of the of the
> memory to be allocated, like where is this memory physically located or what kind of
> shareable handles should be available. If the allocation is successful the function
> returns with a ``hipSuccess`` and the ``hipMemGenericAllocationHandle_t`` will represent
> a valid physical memory allocation. It is important to note that the size has to be
> aligned with the granularity appropriate for the properties of the allocation. You can
> use the ``hipMemGetAllocationGranularity`` function to query this information.

Reviewer 2: ```suggestion
This function accepts the size of the buffer, an ``unsigned long long`` variable for the flags, and a
``hipMemAllocationProp`` variable. ``hipMemAllocationProp`` contains the properties of the 
memory to be allocated, such as where the memory is physically located and what kind of
shareable handles are available. If the allocation is successful, the function
returns a value of ``hipSuccess``, with ``hipMemGenericAllocationHandle_t`` representing
a valid physical memory allocation. The size has to be
aligned with the granularity appropriate for the properties of the allocation. You can
use the ``hipMemGetAllocationGranularity`` function to determine the correct granularity.
```

Reviewer 3: Updated

> .. code-block::cpp

Reviewer 2: I don't see any of the code-blocks in the draft document. I think the syntax is wrong and is missing a space before "cpp"

Reviewer 3: Fixed.

    size_t granularity = 0;
    hipMemGenericAllocationHandle_t allocHandle;
    hipMemAllocationProp prop = {};
    prop.type = HIP_MEM_ALLOCATION_TYPE_PINNED;
    prop.location.type = HIP_MEM_LOCATION_TYPE_DEVICE;
    prop.location.id = currentDev;
    hipMemGetAllocationGranularity(&granularity, &prop, HIP_MEM_ALLOC_GRANULARITY_MINIMUM);
    padded_size = ROUND_UP(size, granularity);
    hipMemCreate(&allocHandle, padded_size, &prop, 0);

Reserve virtual address range
-------------------------------------------------------------------------------

> Once you have a physical memory allocation you need to map it, before you can use it and
> you need a virtual address to map to. To reserve a virtual memory range use the
> ``hipMemAddressReserve`` function, which needs the same size as the physical memory
> allocate previously. Afterwards you can map physical memory allocation to the newly
> acquired virtual memory address range with the ``hipMemMap`` function. Mapping means that
> the physical allocation will be available through the virtual address range it is mapped
> to.

Reviewer 2: ```suggestion
After you have acquired the allocation of physical memory, you must map it before you can use it. To do so,
you need a virtual address to map it to.  Mapping means that the physical memory allocation is available by means of the virtual address range it is mapped to. To reserve a virtual memory range, use the
``hipMemAddressReserve`` function. The size of the virtual memory must match the amount of physical memory previously allocated. You can then map the physical memory allocation to the newly-acquired virtual memory address range using the ``hipMemMap`` function.
```

Reviewer 3: Updated

> .. code-block::cpp

Reviewer 2: Same problem with this code block. They all have this problem.

Reviewer 3: Fixed.

    hipMemAddressReserve(&ptr, padded_size, 0, 0, 0);
    hipMemMap(ptr, padded_size, 0, allocHandle, 0);

Set memory access
-------------------------------------------------------------------------------

> Finally you have to enable access to the memory. This is achieved by the
> ``hipMemSetAccess`` function. It expects the pointer to the virtual memory, the size and
> a ``hipMemAccessDesc`` descriptor as parameters.

Reviewer 2: ```suggestion
Finally, you have to enable access to the memory. This is achieved by using the
``hipMemSetAccess`` function. It accepts the pointer to the virtual memory, the size, and
a ``hipMemAccessDesc`` descriptor as parameters.
```

Reviewer 3: Updated

.. code-block::cpp

    hipMemAccessDesc accessDesc = {};
    accessDesc.location.type = HIP_MEM_LOCATION_TYPE_DEVICE;
    accessDesc.location.id = currentDev;
    accessDesc.flags = HIP_MEM_ACCESS_FLAGS_PROT_READWRITE;
    hipMemSetAccess(ptr, padded_size, &accessDesc, 1);

> At this point you have your memory allocated, mapped and enabled for use. You can read
> and write to it, just like you would with the old C style allocations. 

Reviewer 2: ```suggestion
At this point the memory is allocated, mapped, and ready for use. You can read
and write to it, just like you would a C style memory allocation. 
```

Reviewer 3: Updated

Free virtual memory
-------------------------------------------------------------------------------

> To free the memory allocated like this, there are corresponding free functions. To unmap
> the memory use ``hipMemUnmap``, to release the virtual address range use
> ``hipMemAddressFree`` and finally to release the physical memory use ``hipMemRelease``.

Reviewer 2: ```suggestion
To free the memory allocated in this manner, use the corresponding free functions. To unmap
the memory, use ``hipMemUnmap``. To release the virtual address range, use
``hipMemAddressFree``.  Finally, to release the physical memory, use ``hipMemRelease``.
```

Reviewer 3: Updated

.. code-block::cpp

    hipMemUnmap(ptr, size);
    hipMemRelease(allocHandle);
    hipMemAddressFree(ptr, size);

.. _usage_virtual_memory:
Usage
===============================================================================

Increase allocation size
-------------------------------------------------------------------------------

> This can solve the growing an allocation problem easily. The ``hipMemAddressReserve`` has
> a parameter, that is the requested starting address of the virtual memory. This allows
> you to have a continuous virtual address space, without worrying about the underlying
> physical allocation.

Reviewer 2: ```suggestion
The ``hipMemAddressReserve` function allows you to increase the amount of pre-allocated memory. This function
accepts a parameter representing the requested starting address of the virtual memory. This allows
you to have a continuous virtual address space without worrying about the underlying
physical allocation.
```

Reviewer 3: Updated

.. code-block::cpp

    hipMemAddressReserve(&new_ptr, (new_size - padded_size), 0, ptr + padded_size, 0);
    hipMemMap(new_ptr, (new_size - padded_size), 0, newAllocHandle, 0);
    hipMemSetAccess(new_ptr, (new_size - padded_size), &accessDesc, 1);

> In the code above we assume that ``hipMemAddressReserve`` was able to reserve the memory
> address at the specified location. This isn't guaranteed to be true, so in your
> application check the ``new_ptr`` if you rely on its value being a specific virtual
> address. 

Reviewer 2: ```suggestion
The code sample above assumes that ``hipMemAddressReserve`` was able to reserve the memory
address at the specified location. However, this isn't guaranteed to be true, so you should validate that
``new_ptr`` points to a specific virtual address before using it. 
```

Reviewer 3: Updated

Avoid synchronization
-------------------------------------------------------------------------------

> An other major advantage of virtual memory management is the lack of synchronization
> during the release operations. If you have multiple streams running parallel and you call
> ``hipFree`` it'll synchronize the device, which causes worse resource usage and
> performance.

Reviewer 2: ```suggestion
Another major advantage of virtual memory management is the lack of synchronization
when memory is released. If you call ``hipFree`` when you have multiple streams running in parallel,
it synchronizes the device. This causes worse resource usage and performance.
```

Reviewer 3: Updated

Enable peer access
-------------------------------------------------------------------------------

> In multi-GPU environment you can map the device memory of one GPU to another. This

Reviewer 2: ```suggestion
In a multi-GPU environment, you can map the device memory of one GPU to another. This
```

Reviewer 3: Updated

> feature works with the traditional memory management system as well, but isn't as

Reviewer 2: ```suggestion
feature also works with the traditional memory management system, but isn't as
```

Reviewer 3: Updated

> scalable as with virtual memory. When the memory is allocated with ``hipMalloc`` peer
> access is enabled with ``hipDeviceEnablePeerAccess``. This enables access between two
> devices which means, that every ``hipMalloc`` will be slower to provide the checks and
> the mapping between the devices. Using virtual memory management peer access is enabled
> with the ``hipMemSetAccess``, which means you have a more fine control over what is
> shared. This results in no performance impact on memory allocation and more control on
> what memory buffers are shared with which devices.

Reviewer 2: ```suggestion
scalable as with virtual memory. When memory is allocated with ``hipMalloc``, ``hipDeviceEnablePeerAccess`` is
used to enable peer access. This function enables access between two
devices, but it means that every call to ``hipMalloc`` takes more time to perform the checks and
the mapping between the devices. When using virtual memory management, peer access is enabled
by ``hipMemSetAccess``, which provides a finer level of control over what is
shared. This has no performance impact on memory allocation and gives you more control over
what memory buffers are shared with which devices.
```

Reviewer 3: Updated

.. meta::
  :description: This chapter describes introduces Virtual Memory (VM) and shows
                how to use it in AMD HIP.
  :keywords: AMD, ROCm, HIP, CUDA, virtual memory, virtual, memory, UM, APU

.. _virtual_memory:

*******************************************************************************
HIP Virtual Memory Management
*******************************************************************************

Memory management is very important, when creating high performance applications in the
HIP ecosystem. Both memory allocation and copy can cause bottlenecks in your programs,

> which is a big problem specifically when these operations aren't necessary.

Reviewer 4: ```suggestion
Managing memory management is important when creating high-performance applications in the HIP ecosystem. Both memory allocation and copy can result in bottlenecks, which can significantly impact performance. 
```
